src/uart.rs

```diff
@@ -665,6 +820,222 @@ impl<'d> core::fmt::Write for UartTxDriver<'d> {
+    port: P,
+    _p: PhantomData<&'d mut ()>,
+}
+
+impl<'d, P: sealed::AsPort> Port<'d, P> {
+    fn port(&self) -> uart_port_t {
+        self.port.port()
+    }
+
+    fn stop_bits(&self) -> Result<config::StopBits, EspError> {
+        let mut stop_bits: uart_stop_bits_t = 0;
+        esp_result!(
+            unsafe { uart_get_stop_bits(self.port(), &mut stop_bits) },
+            stop_bits.into()
+        )
+    }
+
+    fn data_bits(&self) -> Result<config::DataBits, EspError> {
+        let mut data_bits: uart_word_length_t = 0;
+        esp_result!(
+            unsafe { uart_get_word_length(self.port(), &mut data_bits) },
+            data_bits.into()
+        )
+    }
+
+    fn parity(&self) -> Result<config::Parity, EspError> {
+        let mut parity: uart_parity_t = 0;
+        esp_result!(
+            unsafe { uart_get_parity(self.port(), &mut parity) },
+            parity.into()
+        )
+    }
+
+    fn baudrate(&self) -> Result<Hertz, EspError> {
+        let mut baudrate: u32 = 0;
+        esp_result!(
+            unsafe { uart_get_baudrate(self.port(), &mut baudrate) },
+            baudrate.into()
+        )
+    }
+}
+
+impl<'d> Port<'d, Owned> {
+    fn borrow(&mut self) -> Port<'d, Borrowed> {
+        Port {
+            port: Borrowed(self.port() as _),
+            _p: PhantomData,
+        }
+    }
+
+    fn change_stop_bits(&mut self, stop_bits: config::StopBits) -> Result<(), EspError> {
+        esp!(unsafe { uart_set_stop_bits(self.port(), stop_bits.into()) })
+    }
+
+    fn change_data_bits(&mut self, data_bits: config::DataBits) -> Result<(), EspError> {
+        esp!(unsafe { uart_set_word_length(self.port(), data_bits.into()) })
+    }
+
+    fn change_parity(&mut self, parity: config::Parity) -> Result<(), EspError> {
+        esp!(unsafe { uart_set_parity(self.port(), parity.into()) })
+    }
+
+    fn change_baudrate<T: Into<Hertz> + Copy>(&mut self, baudrate: T) -> Result<(), EspError> {
+        esp!(unsafe { uart_set_baudrate(self.port(), baudrate.into().into()) })
+    }
+}
+
+mod sealed {
+    pub trait AsPort {
+        fn port(&self) -> esp_idf_sys::uart_port_t;
+    }
+}
+
+pub struct Owned(u8);
+
+impl sealed::AsPort for Owned {
+    fn port(&self) -> uart_port_t {
+        self.0 as _
+    }
+}
+
+impl Drop for Owned {
+    fn drop(&mut self) {
+        esp!(unsafe { uart_driver_delete(self.0 as _) }).unwrap()
+    }
+}
+
+pub struct Borrowed(u8);
+
+impl sealed::AsPort for Borrowed {
+    fn port(&self) -> uart_port_t {
+        self.0 as _
+    }
+}
+
+#[cfg(feature = "alloc")]
+mod shared {
+    use super::*;
+
+    extern crate alloc;
+    use alloc::sync::Arc;
+
+    pub struct Shared(Arc<Owned>);
+
+    impl<'d> UartDriver<'d> {
+        /// Split the serial driver in separate TX and RX drivers.
+        ///
+        /// Unlike [`split`], the halves are owned, at the cost of an allocation.
+        pub fn into_split(self) -> (UartTxDriver<'d, Shared>, UartRxDriver<'d, Shared>) {
```

Comment:
I don't think the `Shared` concept would be necessary at all, if you abandon the extra generics and embrace the `{ port: u8, owned: bool }` idea. In that case, you really only need to properly manage the lifetimes (I think). In other words:
* `fn split<'a>(&'a mut self) -> (UartTxDriver<'a>, UartRxDriver<'a>)` // Notice how the TX/RX drivers take a `mut` ref on the main driver, with a different lifetime from `'d`
* `fn into_split(self) -> (UartTxDriver<'d>, UartRxDriver<'d>)` // Notice how the TX/RX drivers "inherit" the `'d` lifetime of the main driver